src/transformers/models/doge/modular_doge.py

```diff
@@ -0,0 +1,1172 @@
+        self.rms_norm_eps = rms_norm_eps
+        self.use_cache = use_cache
+
+        self.max_position_embeddings = max_position_embeddings
+        self.rope_theta = rope_theta
+        self.rope_scaling = rope_scaling
+        self.num_attention_heads = num_attention_heads
+        self.num_key_value_heads = num_key_value_heads
+        self.attention_dropout = attention_dropout
+        self.dynamic_mask_ratio = dynamic_mask_ratio
+        self.is_moe = is_moe
+        self.num_cdmoe_experts = num_cdmoe_experts
+        self.num_cdmoe_heads = num_cdmoe_heads
+        self.num_cdmoe_experts_per_head = num_cdmoe_experts_per_head
+        self.expert_retrieval_size = expert_retrieval_size
+
+        # Validate the correctness of rotary position embeddings parameters
+        # BC: if there is a 'type' field, copy it it to 'rope_type'.
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+
+        # for backward compatibility
+        if num_key_value_heads is None:
+            self.num_key_value_heads = num_attention_heads
+
+        super().__init__(
+            bos_token_id=bos_token_id,
+            eos_token_id=eos_token_id,
+            pad_token_id=pad_token_id,
+            tie_word_embeddings=tie_word_embeddings,
+            **kwargs,
+        )
+
+
+class RMSNorm(LlamaRMSNorm):
+    def __init__(self, hidden_size, eps=1e-6):
+        """
+        RMSNorm is equivalent to T5LayerNorm
+        """
+        super().__init__(hidden_size, eps)
+
+
+class Residual(nn.Module):
+    def __init__(self, hidden_size):
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(hidden_size))
+
+    def forward(self, residual_states, hidden_states):
+        return self.weight * residual_states + hidden_states
+
+    def extra_repr(self):
+        return f"{tuple(self.weight.shape)}"
+
+
+class RotaryEmbedding(LlamaRotaryEmbedding):
+    def __init__(self, config: Optional[DogeConfig] = None, device=None):
+        super().__init__(config, device)
+
+
+class DogeDynamicMaskAttention(nn.Module):
+    """Dynamic Mask Attention from 'Wonderful Matrices' paper."""
+
+    def __init__(self, config: DogeConfig, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.head_dim = config.hidden_size // config.num_attention_heads
+        self.num_key_value_groups = config.num_attention_heads // config.num_key_value_heads
+        self.scaling = self.head_dim**-0.5
+        self.attention_dropout = config.attention_dropout
+        self.dynamic_mask_ratio = config.dynamic_mask_ratio
+
+        self.ALL_ATTENTION_FUNCTIONS = {
+            "eager": self.eager_attention_forward,
+            "flex_attention": self.flex_attention_forward,
+            "sdpa": self.sdpa_attention_forward,
+        }
```

Comment:
if it is only in self, it's gonna be hard to modify outside the scope of this model only! If someone wants to add "my_custom_function" he is gonna have a hard time!